File: qt/gen_qmimedata.cpp
// +build ignore

#include <QByteArray>
#include <QList>
#include <QMetaObject>
#include <QMetaType>
#include <QMimeData>
#include <QObject>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QUrl>
#include <QVariant>
#include <qmimedata.h>
#include "gen_qmimedata.h"
class MiqtVirtualQMimeData : public virtual QMimeData {
public:
MiqtVirtualQMimeData(): QMimeData() {};

virtual ~MiqtVirtualQMimeData() = default;
// cgo.Handle value for overwritten implementation
	intptr_t handle__MetaObject = 0;
// Subclass to allow providing a Go implementation
	virtual const QMetaObject* metaObject() const override {
if (handle__MetaObject == 0) {
return QMimeData::metaObject();
}
QMetaObject* callback_return_value = miqt_exec_callback_QMimeData_MetaObject(const_cast<MiqtVirtualQMimeData*>(this), handle__MetaObject);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	QMetaObject* virtualbase_MetaObject() const {

		return (QMetaObject*) QMimeData::metaObject();

	}
// cgo.Handle value for overwritten implementation
	intptr_t handle__Metacast = 0;
// Subclass to allow providing a Go implementation
	virtual void* qt_metacast(const char* param1) override {
if (handle__Metacast == 0) {
return QMimeData::qt_metacast(param1);
}
const char* sigval1 = (const char*) param1;

		void* callback_return_value = miqt_exec_callback_QMimeData_Metacast(this, handle__Metacast, sigval1);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	void* virtualbase_Metacast(const char* param1) {

		return QMimeData::qt_metacast(param1);

	}
};
QMimeData* QMimeData_new() {
return new MiqtVirtualQMimeData();
}
void QMimeData_virtbase(QMimeData* src
, QObject** outptr_QObject
) {
*outptr_QObject = static_cast<QObject*>(src);
}
QMetaObject* QMimeData_MetaObject(const QMimeData* self) {
	return (QMetaObject*) self->metaObject();
}
void* QMimeData_Metacast(QMimeData* self, const char* param1) {
	return self->qt_metacast(param1);
}
struct miqt_string QMimeData_Tr(const char* s) {
	QString _ret = QMimeData::tr(s);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
struct miqt_array /* of QUrl* */  QMimeData_Urls(const QMimeData* self) {
	QList<QUrl> _ret = self->urls();
	// Convert QList<> from C++ memory to manually-managed C memory
	QUrl** _arr = static_cast<QUrl**>(malloc(sizeof(QUrl*) * _ret.length()));
	for (size_t i = 0, e = _ret.length(); i < e; ++i) {
		_arr[i] = new QUrl(_ret[i]);
	}
	struct miqt_array _out;
	_out.len = _ret.length();
	_out.data = static_cast<void*>(_arr);
	return _out;
}
void QMimeData_SetUrls(QMimeData* self, struct miqt_array /* of QUrl* */  urls) {
	QList<QUrl> urls_QList;
	urls_QList.reserve(urls.len);
	QUrl** urls_arr = static_cast<QUrl**>(urls.data);
	for(size_t i = 0; i < urls.len; ++i) {
		urls_QList.push_back(*(urls_arr[i]));
	}
	self->setUrls(urls_QList);
}
bool QMimeData_HasUrls(const QMimeData* self) {
	return self->hasUrls();
}
struct miqt_string QMimeData_Text(const QMimeData* self) {
	QString _ret = self->text();
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
void QMimeData_SetText(QMimeData* self, struct miqt_string text) {
	QString text_QString = QString::fromUtf8(text.data, text.len);
	self->setText(text_QString);
}
bool QMimeData_HasText(const QMimeData* self) {
	return self->hasText();
}
struct miqt_string QMimeData_Html(const QMimeData* self) {
	QString _ret = self->html();
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
void QMimeData_SetHtml(QMimeData* self, struct miqt_string html) {
	QString html_QString = QString::fromUtf8(html.data, html.len);
	self->setHtml(html_QString);
}
bool QMimeData_HasHtml(const QMimeData* self) {
	return self->hasHtml();
}
QVariant* QMimeData_ImageData(const QMimeData* self) {
	return new QVariant(self->imageData());
}
void QMimeData_SetImageData(QMimeData* self, QVariant* image) {
	self->setImageData(*image);
}
bool QMimeData_HasImage(const QMimeData* self) {
	return self->hasImage();
}
QVariant* QMimeData_ColorData(const QMimeData* self) {
	return new QVariant(self->colorData());
}
void QMimeData_SetColorData(QMimeData* self, QVariant* color) {
	self->setColorData(*color);
}
bool QMimeData_HasColor(const QMimeData* self) {
	return self->hasColor();
}
struct miqt_string QMimeData_Data(const QMimeData* self, struct miqt_string mimetype) {
	QString mimetype_QString = QString::fromUtf8(mimetype.data, mimetype.len);
	QByteArray _qb = self->data(mimetype_QString);
	struct miqt_string _ms;
	_ms.len = _qb.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _qb.data(), _ms.len);
	return _ms;
}
void QMimeData_SetData(QMimeData* self, struct miqt_string mimetype, struct miqt_string data) {
	QString mimetype_QString = QString::fromUtf8(mimetype.data, mimetype.len);
	QByteArray data_QByteArray(data.data, data.len);
	self->setData(mimetype_QString, data_QByteArray);
}
void QMimeData_RemoveFormat(QMimeData* self, struct miqt_string mimetype) {
	QString mimetype_QString = QString::fromUtf8(mimetype.data, mimetype.len);
	self->removeFormat(mimetype_QString);
}
bool QMimeData_HasFormat(const QMimeData* self, struct miqt_string mimetype) {
	QString mimetype_QString = QString::fromUtf8(mimetype.data, mimetype.len);
	return self->hasFormat(mimetype_QString);
}
struct miqt_array /* of struct miqt_string */  QMimeData_Formats(const QMimeData* self) {
	QStringList _ret = self->formats();
	// Convert QList<> from C++ memory to manually-managed C memory
	struct miqt_string* _arr = static_cast<struct miqt_string*>(malloc(sizeof(struct miqt_string) * _ret.length()));
	for (size_t i = 0, e = _ret.length(); i < e; ++i) {
		QString _lv_ret = _ret[i];
		// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
		QByteArray _lv_b = _lv_ret.toUtf8();
		struct miqt_string _lv_ms;
		_lv_ms.len = _lv_b.length();
		_lv_ms.data = static_cast<char*>(malloc(_lv_ms.len));
		memcpy(_lv_ms.data, _lv_b.data(), _lv_ms.len);
		_arr[i] = _lv_ms;
	}
	struct miqt_array _out;
	_out.len = _ret.length();
	_out.data = static_cast<void*>(_arr);
	return _out;
}
void QMimeData_Clear(QMimeData* self) {
	self->clear();
}
struct miqt_string QMimeData_Tr2(const char* s, const char* c) {
	QString _ret = QMimeData::tr(s, c);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
struct miqt_string QMimeData_Tr3(const char* s, const char* c, int n) {
	QString _ret = QMimeData::tr(s, c, static_cast<int>(n));
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
void QMimeData_override_virtual_MetaObject(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQMimeData*>( (QMimeData*)(self) )->handle__MetaObject = slot;
}
QMetaObject* QMimeData_virtualbase_MetaObject(const void* self) {
	return ( (const MiqtVirtualQMimeData*)(self) )->virtualbase_MetaObject();
}
void QMimeData_override_virtual_Metacast(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQMimeData*>( (QMimeData*)(self) )->handle__Metacast = slot;
}
void* QMimeData_virtualbase_Metacast(void* self, const char* param1) {
	return ( (MiqtVirtualQMimeData*)(self) )->virtualbase_Metacast(param1);
}
void QMimeData_Delete(QMimeData* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<MiqtVirtualQMimeData*>( self );
	} else {
		delete self;
	}
}

File: qt/gen_qmimedata.go
package qt

import (
	"unsafe"
)

type QMimeData struct {
	h          uintptr
	isSubclass bool
}

// NewQMimeData constructs a new QMimeData object.
func NewQMimeData() *QMimeData {

	ret := newQMimeData(QMimeData_new())
	ret.isSubclass = true
	return ret
}

func (this *QMimeData) MetaObject() *QMetaObject {
	return newQMetaObject(QMimeData_MetaObject(this.h))
}

func (this *QMimeData) Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))
	return (unsafe.Pointer)(QMimeData_Metacast(this.h, param1_Cstring))
}

func QMimeData_Tr(s string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	var _ms struct_miqt_string = QMimeData_Tr(s_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QMimeData) Urls() []QUrl {
	var _ma struct_miqt_array = QMimeData_Urls(this.h)
	_ret := make([]QUrl, int(_ma.len))
	_outCast := (*[0xffff]*QUrl)(unsafe.Pointer(_ma.data)) // hey ya
	for i := 0; i < int(_ma.len); i++ {
		_lv_goptr := newQUrl(_outCast[i])
		_lv_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
		_ret[i] = *_lv_goptr
	}
	return _ret
}

func (this *QMimeData) SetUrls(urls []QUrl) {
	urls_CArray := (*[0xffff]*QUrl)(malloc(size_t(8 * len(urls))))
	defer free(unsafe.Pointer(urls_CArray))
	for i := range urls {
		urls_CArray[i] = urls[i].cPointer()
	}
	urls_ma := struct_miqt_array{len: size_t(len(urls)), data: unsafe.Pointer(urls_CArray)}
	QMimeData_SetUrls(this.h, urls_ma)
}

func (this *QMimeData) HasUrls() bool {
	return (bool)(QMimeData_HasUrls(this.h))
}

func (this *QMimeData) Text() string {
	var _ms struct_miqt_string = QMimeData_Text(this.h)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QMimeData) SetText(text string) {
	text_ms := struct_miqt_string{}
	text_ms.data = CString(text)
	text_ms.len = size_t(len(text))
	defer free(unsafe.Pointer(text_ms.data))
	QMimeData_SetText(this.h, text_ms)
}

func (this *QMimeData) HasText() bool {
	return (bool)(QMimeData_HasText(this.h))
}

func (this *QMimeData) Html() string {
	var _ms struct_miqt_string = QMimeData_Html(this.h)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QMimeData) SetHtml(html string) {
	html_ms := struct_miqt_string{}
	html_ms.data = CString(html)
	html_ms.len = size_t(len(html))
	defer free(unsafe.Pointer(html_ms.data))
	QMimeData_SetHtml(this.h, html_ms)
}

func (this *QMimeData) HasHtml() bool {
	return (bool)(QMimeData_HasHtml(this.h))
}

func (this *QMimeData) ImageData() *QVariant {
	_goptr := newQVariant(QMimeData_ImageData(this.h))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QMimeData) SetImageData(image *QVariant) {
	QMimeData_SetImageData(this.h, image.cPointer())
}

func (this *QMimeData) HasImage() bool {
	return (bool)(QMimeData_HasImage(this.h))
}

func (this *QMimeData) ColorData() *QVariant {
	_goptr := newQVariant(QMimeData_ColorData(this.h))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QMimeData) SetColorData(color *QVariant) {
	QMimeData_SetColorData(this.h, color.cPointer())
}

func (this *QMimeData) HasColor() bool {
	return (bool)(QMimeData_HasColor(this.h))
}

func (this *QMimeData) Data(mimetype string) []byte {
	mimetype_ms := struct_miqt_string{}
	mimetype_ms.data = CString(mimetype)
	mimetype_ms.len = size_t(len(mimetype))
	defer free(unsafe.Pointer(mimetype_ms.data))
	var _bytearray struct_miqt_string = QMimeData_Data(this.h, mimetype_ms)
	_ret := GoBytes(unsafe.Pointer(_bytearray.data), int(int64(_bytearray.len)))
	free(unsafe.Pointer(_bytearray.data))
	return _ret
}

func (this *QMimeData) SetData(mimetype string, data []byte) {
	mimetype_ms := struct_miqt_string{}
	mimetype_ms.data = CString(mimetype)
	mimetype_ms.len = size_t(len(mimetype))
	defer free(unsafe.Pointer(mimetype_ms.data))
	data_alias := struct_miqt_string{}
	data_alias.data = (char)(unsafe.Pointer(&data[0]))
	data_alias.len = size_t(len(data))
	QMimeData_SetData(this.h, mimetype_ms, data_alias)
}

func (this *QMimeData) RemoveFormat(mimetype string) {
	mimetype_ms := struct_miqt_string{}
	mimetype_ms.data = CString(mimetype)
	mimetype_ms.len = size_t(len(mimetype))
	defer free(unsafe.Pointer(mimetype_ms.data))
	QMimeData_RemoveFormat(this.h, mimetype_ms)
}

func (this *QMimeData) HasFormat(mimetype string) bool {
	mimetype_ms := struct_miqt_string{}
	mimetype_ms.data = CString(mimetype)
	mimetype_ms.len = size_t(len(mimetype))
	defer free(unsafe.Pointer(mimetype_ms.data))
	return (bool)(QMimeData_HasFormat(this.h, mimetype_ms))
}

func (this *QMimeData) Formats() []string {
	var _ma struct_miqt_array = QMimeData_Formats(this.h)
	_ret := make([]string, int(_ma.len))
	_outCast := (*[0xffff]struct_miqt_string)(unsafe.Pointer(_ma.data)) // hey ya
	for i := 0; i < int(_ma.len); i++ {
		var _lv_ms struct_miqt_string = _outCast[i]
		_lv_ret := GoStringN(_lv_ms.data, int(int64(_lv_ms.len)))
		free(unsafe.Pointer(_lv_ms.data))
		_ret[i] = _lv_ret
	}
	return _ret
}

func (this *QMimeData) Clear() {
	QMimeData_Clear(this.h)
}

func QMimeData_Tr2(s string, c string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QMimeData_Tr2(s_Cstring, c_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func QMimeData_Tr3(s string, c string, n int) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QMimeData_Tr3(s_Cstring, c_Cstring, (int)(n))
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QMimeData) callVirtualBase_HasFormat(mimetype string) bool {
	mimetype_ms := struct_miqt_string{}
	mimetype_ms.data = CString(mimetype)
	mimetype_ms.len = size_t(len(mimetype))
	defer free(unsafe.Pointer(mimetype_ms.data))

	return (bool)(QMimeData_virtualbase_HasFormat(unsafe.Pointer(this.h), mimetype_ms))

}
func (this *QMimeData) OnHasFormat(slot func(super func(mimetype string) bool, mimetype string) bool) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QMimeData_override_virtual_HasFormat(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QMimeData_HasFormat
func miqt_exec_callback_QMimeData_HasFormat(self QMimeData, cb intptr_t, mimetype struct_miqt_string) bool {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func(mimetype string) bool, mimetype string) bool)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	var mimetype_ms struct_miqt_string = mimetype
	mimetype_ret := GoStringN(mimetype_ms.data, int(int64(mimetype_ms.len)))
	free(unsafe.Pointer(mimetype_ms.data))
	slotval1 := mimetype_ret

	virtualReturn := gofunc((&QMimeData{h: self}).callVirtualBase_HasFormat, slotval1)

	return (bool)(virtualReturn)

}

func (this *QMimeData) callVirtualBase_Formats() []string {

	var _ma struct_miqt_array = QMimeData_virtualbase_Formats(unsafe.Pointer(this.h))
	_ret := make([]string, int(_ma.len))
	_outCast := (*[0xffff]struct_miqt_string)(unsafe.Pointer(_ma.data)) // hey ya
	for i := 0; i < int(_ma.len); i++ {
		var _lv_ms struct_miqt_string = _outCast[i]
		_lv_ret := GoStringN(_lv_ms.data, int(int64(_lv_ms.len)))
		free(unsafe.Pointer(_lv_ms.data))
		_ret[i] = _lv_ret
	}
	return _ret

}
func (this *QMimeData) OnFormats(slot func(super func() []string) []string) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QMimeData_override_virtual_Formats(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QMimeData_Formats
func miqt_exec_callback_QMimeData_Formats(self QMimeData, cb intptr_t) struct_miqt_array {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func() []string) []string)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	virtualReturn := gofunc((&QMimeData{h: self}).callVirtualBase_Formats)
	virtualReturn_CArray := (*[0xffff]struct_miqt_string)(malloc(size_t(int(unsafe.Sizeof(struct_miqt_string{})) * len(virtualReturn))))
	defer free(unsafe.Pointer(virtualReturn_CArray))
	for i := range virtualReturn {
		virtualReturn_i_ms := struct_miqt_string{}
		virtualReturn_i_ms.data = CString(virtualReturn[i])
		virtualReturn_i_ms.len = size_t(len(virtualReturn[i]))
		defer free(unsafe.Pointer(virtualReturn_i_ms.data))
		virtualReturn_CArray[i] = virtualReturn_i_ms
	}
	virtualReturn_ma := struct_miqt_array{len: size_t(len(virtualReturn)), data: unsafe.Pointer(virtualReturn_CArray)}

	return virtualReturn_ma

}

func (this *QMimeData) callVirtualBase_RetrieveData(mimetype string, preferredType QMetaType) *QVariant {
	mimetype_ms := struct_miqt_string{}
	mimetype_ms.data = CString(mimetype)
	mimetype_ms.len = size_t(len(mimetype))
	defer free(unsafe.Pointer(mimetype_ms.data))

	_goptr := newQVariant(QMimeData_virtualbase_RetrieveData(unsafe.Pointer(this.h), mimetype_ms, preferredType.cPointer()))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr

}
func (this *QMimeData) OnRetrieveData(slot func(super func(mimetype string, preferredType QMetaType) *QVariant, mimetype string, preferredType QMetaType) *QVariant) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QMimeData_override_virtual_RetrieveData(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QMimeData_RetrieveData
func miqt_exec_callback_QMimeData_RetrieveData(self QMimeData, cb intptr_t, mimetype struct_miqt_string, preferredType *QMetaType) *QVariant {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func(mimetype string, preferredType QMetaType) *QVariant, mimetype string, preferredType QMetaType) *QVariant)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	var mimetype_ms struct_miqt_string = mimetype
	mimetype_ret := GoStringN(mimetype_ms.data, int(int64(mimetype_ms.len)))
	free(unsafe.Pointer(mimetype_ms.data))
	slotval1 := mimetype_ret
	preferredType_goptr := newQMetaType(preferredType)
	preferredType_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	slotval2 := *preferredType_goptr

	virtualReturn := gofunc((&QMimeData{h: self}).callVirtualBase_RetrieveData, slotval1, slotval2)

	return virtualReturn.cPointer()

}

func (this *QMimeData) callVirtualBase_Event(event *QEvent) bool {

	return (bool)(QMimeData_virtualbase_Event(unsafe.Pointer(this.h), event.cPointer()))

}
func (this *QMimeData) OnEvent(slot func(super func(event *QEvent) bool, event *QEvent) bool) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QMimeData_override_virtual_Event(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QMimeData_Event
func miqt_exec_callback_QMimeData_Event(self QMimeData, cb intptr_t, event *QEvent) bool {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func(event *QEvent) bool, event *QEvent) bool)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	slotval1 := newQEvent(event)

	virtualReturn := gofunc((&QMimeData{h: self}).callVirtualBase_Event, slotval1)

	return (bool)(virtualReturn)

}

func (this *QMimeData) callVirtualBase_EventFilter(watched *QObject, event *QEvent) bool {

	return (bool)(QMimeData_virtualbase_EventFilter(unsafe.Pointer(this.h), watched.cPointer(), event.cPointer()))

}
func (this *QMimeData) OnEventFilter(slot func(super func(watched *QObject, event *QEvent) bool, watched *QObject, event *QEvent) bool) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QMimeData_override_virtual_EventFilter(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QMimeData_EventFilter
func miqt_exec_callback_QMimeData_EventFilter(self QMimeData, cb intptr_t, watched *QObject, event *QEvent) bool {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func(watched *QObject, event *QEvent) bool, watched *QObject, event *QEvent) bool)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	slotval1 := newQObject(watched)

	slotval2 := newQEvent(event)

	virtualReturn := gofunc((&QMimeData{h: self}).callVirtualBase_EventFilter, slotval1, slotval2)

	return (bool)(virtualReturn)

}

func (this *QMimeData) callVirtualBase_TimerEvent(event *QTimerEvent) {

	QMimeData_virtualbase_TimerEvent(unsafe.Pointer(this.h), event.cPointer())

}
func (this *QMimeData) OnTimerEvent(slot func(super func(event *QTimerEvent), event *QTimerEvent)) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QMimeData_override_virtual_TimerEvent(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QMimeData_TimerEvent
func miqt_exec_callback_QMimeData_TimerEvent(self QMimeData, cb intptr_t, event *QTimerEvent) {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func(event *QTimerEvent), event *QTimerEvent))
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	slotval1 := newQTimerEvent(event)

	gofunc((&QMimeData{h: self}).callVirtualBase_TimerEvent, slotval1)

}

func (this *QMimeData) callVirtualBase_ChildEvent(event *QChildEvent) {

	QMimeData_virtualbase_ChildEvent(unsafe.Pointer(this.h), event.cPointer())

}
func (this *QMimeData) OnChildEvent(slot func(super func(event *QChildEvent), event *QChildEvent)) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QMimeData_override_virtual_ChildEvent(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QMimeData_ChildEvent
func miqt_exec_callback_QMimeData_ChildEvent(self QMimeData, cb intptr_t, event *QChildEvent) {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func(event *QChildEvent), event *QChildEvent))
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	slotval1 := newQChildEvent(event)

	gofunc((&QMimeData{h: self}).callVirtualBase_ChildEvent, slotval1)

}

func (this *QMimeData) callVirtualBase_CustomEvent(event *QEvent) {

	QMimeData_virtualbase_CustomEvent(unsafe.Pointer(this.h), event.cPointer())

}
func (this *QMimeData) OnCustomEvent(slot func(super func(event *QEvent), event *QEvent)) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QMimeData_override_virtual_CustomEvent(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QMimeData_CustomEvent
func miqt_exec_callback_QMimeData_CustomEvent(self QMimeData, cb intptr_t, event *QEvent) {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func(event *QEvent), event *QEvent))
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	slotval1 := newQEvent(event)

	gofunc((&QMimeData{h: self}).callVirtualBase_CustomEvent, slotval1)

}

func (this *QMimeData) callVirtualBase_ConnectNotify(signal *QMetaMethod) {

	QMimeData_virtualbase_ConnectNotify(unsafe.Pointer(this.h), signal.cPointer())

}
func (this *QMimeData) OnConnectNotify(slot func(super func(signal *QMetaMethod), signal *QMetaMethod)) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QMimeData_override_virtual_ConnectNotify(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QMimeData_ConnectNotify
func miqt_exec_callback_QMimeData_ConnectNotify(self QMimeData, cb intptr_t, signal *QMetaMethod) {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func(signal *QMetaMethod), signal *QMetaMethod))
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	slotval1 := newQMetaMethod(signal)

	gofunc((&QMimeData{h: self}).callVirtualBase_ConnectNotify, slotval1)

}

func (this *QMimeData) callVirtualBase_DisconnectNotify(signal *QMetaMethod) {

	QMimeData_virtualbase_DisconnectNotify(unsafe.Pointer(this.h), signal.cPointer())

}
func (this *QMimeData) OnDisconnectNotify(slot func(super func(signal *QMetaMethod), signal *QMetaMethod)) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QMimeData_override_virtual_DisconnectNotify(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QMimeData_DisconnectNotify
func miqt_exec_callback_QMimeData_DisconnectNotify(self QMimeData, cb intptr_t, signal *QMetaMethod) {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func(signal *QMetaMethod), signal *QMetaMethod))
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	slotval1 := newQMetaMethod(signal)

	gofunc((&QMimeData{h: self}).callVirtualBase_DisconnectNotify, slotval1)

}
